### DISCUSSION TITLE
How to specify links order?

### DISCUSSION THREAD

**Discussion by Jarod42** [premake/premake-core#1989] [Q&A]
For non-msvc compilers (gcc/clang), [order of links might be important](https://stackoverflow.com/questions/45135/why-does-the-order-in-which-libraries-are-linked-sometimes-cause-errors-in-gcc):
if A depends on B which depends on C
we should have something like:
```
gcc -o A.exe src_A.o -lB -lC
```
(several entries is even allowed for cycle).

In premake, we have 3 kinds of link options:
- `links("project")`
- `links("external_lib")`
- `linkoptions("extra option")`

I don't find a way to order `links` against `linkoptions`.
Generators use one block after the others, (and no order is more correct than the other BTW).

I have basically
```
project "my_project_using_llvm"
  kind "StaticLib"
  -- ...
project "app"
  links("my_project_using_llvm")
  linkoptions { "$(shell llvm-config some_args)" }
```
Using either gmake2/codelite action, the link fails.

As work around, I end up with ugly:
```
project "my_project_using_llvm"
  kind "StaticLib"
  -- ...
project "app"
  links("my_project_using_llvm") -- To keep dependency, and duplicate in linkoptions in expected order
  filter { "system:windows" }
    linkoptions{path.join("%{cfg.targetdir}", "my_project_using_llvm.lib")}
  filter { "system:not windows" }
    linkoptions{path.join("%{cfg.targetdir}", "libmy_project_using_llvm.a")}
  filter {}

  linkoptions { "$(shell llvm-config some_args)" }
```

Is there currently a better way?

**Comment by samsinsane** [comment]
Using [`linkgroups`](https://premake.github.io/docs/linkgroups/) is intended to bypass the ordering issue.

**Comment by Jarod42** [comment]
`linkgroups` only applies to `links`.
So it doesn't resolve the issues with `linkoptions`. I try to add also `Wl, --start-group`/`Wl, --end-group` for linksoptions, but issue is between groups, not inside them...
